libraries/ESP8266WiFi/src/ESP8266WiFiMulti.cpp

```diff
@@ -1,267 +1,507 @@
+            DEBUG_WIFI_MULTI("[WIFIM] Scan timeout\n");
+            return WIFI_SCAN_FAILED;
+        }
+
+        // Get scan result
+        scanResult = WiFi.scanComplete();
+    } while (scanResult < 0);
+
+    // Print WiFi scan result
+    printWiFiScan();
+
+    // Return (positive) number of detected WiFi networks
+    return scanResult;
+}
+
+/**
+ * @brief Connect to multiple WiFi's
+ * @param connectTimeoutMs
+ *      WiFi connect timeout in ms
+ * @return
+ *      WiFi conneciton status
+ */
+wl_status_t ESP8266WiFiMulti::connectWiFiMulti(uint32_t connectTimeoutMs)
+{
+    int8_t scanResult;
+    String ssid;
+    int32_t rssi;
+    uint8_t encType;
+    uint8_t *bssid;
+    int32_t channel;
+    bool hidden;
+
+    // Get scan results
+    scanResult = WiFi.scanComplete();
+
+    // Find known WiFi networks
+    uint8_t known[_APlist.size()];
+    uint8_t numNetworks = 0;
+    for (int8_t i = 0; i < scanResult; i++) {
+        // Get network information
+        WiFi.getNetworkInfo(i, ssid, encType, rssi, bssid, channel, hidden);
+
+        // Check if the WiFi network contains an entry in AP list
+        for (auto entry : _APlist) {
+            // Check SSID
+            if (ssid == entry.ssid) {
+                // Known network
+                known[numNetworks++] = i;
+            }
+        }
+    }
+
+    // Sort WiFi networks by RSSI
+    for (int i = 0; i < numNetworks; i++) {
+        for (int j = i + 1; j < numNetworks; j++) {
+            if (WiFi.RSSI(known[j]) > WiFi.RSSI(known[i])) {
+                int8_t tmp;
+
+                // Swap indices
+                tmp = known[i];
+                known[i] = known[j];
+                known[j] = tmp;
+            }
+        }
+    }
+
+    // Print sorted indices
+    DEBUG_WIFI_MULTI("[WIFIM] Sorted indices: ");
+    for (int8_t i = 0; i < numNetworks; i++) {
+        DEBUG_WIFI_MULTI("%d ", known[i]);
+    }
+    DEBUG_WIFI_MULTI("\n");
+
+    // Connect to known WiFi AP's sorted by RSSI
+    for (int8_t i = 0; i < numNetworks; i++) {
+        // Get network information
+        WiFi.getNetworkInfo(known[i], ssid, encType, rssi, bssid, channel, hidden);
+
+        for (auto entry : _APlist) {
+            // Check SSID
+            if (ssid == entry.ssid) {
+                DEBUG_WIFI_MULTI("[WIFIM] Connecting %s\n", ssid);
+
+                // Connect to WiFi
+                WiFi.begin(ssid, entry.passphrase, channel, bssid);
+
+                // Wait for status change
+                if (waitWiFiConnect(connectTimeoutMs) == WL_CONNECTED) {
+                    return WL_CONNECTED;
+                }
+            }
+        }
+    }
+
+    DEBUG_WIFI_MULTI("[WIFIM] Could not connect\n", ssid);
+
+    // Coult not connect to any WiFi network
+    return WL_CONNECT_FAILED;
+}
+
+/**
+ * @brief Print WiFi scan results
+ * @details
+ *      Macro DEBUG_ESP_WIFI and DEBUG_ESP_PORT must be configured
+ */
+void ESP8266WiFiMulti::printWiFiScan()
```

Comment:
This method doesn't depend on anything from ESP8266WiFiMulti. Please remove it from the class and convert it to a standalone function that is private to the cpp (static or inside anonymous namespace)

Reply:
Accepted.